File: current_session/python/151.py
"""
151
"""
class Solution:
    def reverseWords(self, s: str) -> str:
        i = 0
        res = []
        while i < len(s):
            while i < len(s) and s[i] == ' ':
                i += 1
            j = i+1
            while j < len(s) and s[j] != ' ':
                j += 1
            if s[i:j]:
                res += s[i:j],
            i = j+1
        
        # print(res)
        
        return ' '.join(res[::-1])


print(Solution().reverseWords("the sky is blue"))
print(Solution().reverseWords("  hello world  "))
print(Solution().reverseWords("a good   example"))
print(Solution().reverseWords("  Bob    Loves  Alice   "))

File: current_session/c++/151.cpp
#include <iostream>
#include <string>
#include <stack>

using namespace std;

// put every word into a vector
// reverse the order of vector
class Solution {
public:
  string reverseWords(string s) {
    stack<string> st = helper(s);
    string ans;
    while (!st.empty()) {
      string w = st.top();
      st.pop();
      ans = ans + w;
      if (!st.empty()) ans += " ";
    }
    return ans;

  }
private:
  stack<string> helper(string s) {
    int i=0;
    stack<string> wordStack;
    while (i<s.size()) {
      if (s[i]!=' ') {
        int j=i+1;
        while (j<s.size() && s[j] != ' ') ++j;
        wordStack.push(s.substr(i, j-i));
        i = j;
      } else {
        ++i;
      }
    }
    
    return wordStack;
  }
};

int main() {
  string sss = "the sky is blue";
  cout << Solution().reverseWords(sss) << endl;
}